[SYCL] Add support for RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR macro (#675)
intel/llvm: sycl/doc/SYCLPreprocessorMacroses.md

# Overview

> This file describes macroses that are having effect on SYCL compiler and run-time.

Reviewer: ```suggestion
This file describes macros that are having effect on SYCL compiler and run-time.
```

Author: Done.

### RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR

> Spec assumes that SYCL implementation does address space deduction, but we made

Reviewer: ```suggestion
The spec assumes that the SYCL implementation does address space deduction. However,
for our implementation, the deduction is performed in the middle end, where it's
```

an implementation doing address space deduction in the middle end, where it's
hard to provide user friendly diagnositcs.
Due to these problems writing to raw pointers obtained from `constant_ptr` is not
diagnosed now.

> User can enable diagnostics of writing to such pointers via enabling of

Reviewer: ```suggestion
The user can enable diagnostics upon writing to such pointers via enabling the
`RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` macro.
```

> `RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` macro.

Reviewer: The LIT tests don't seem to be covering this. Is this intentional?

Author: Diagnostics will point to SYCL headers, I don't know the way how to validate diagnostics pointing not to test file. If you know, we can probably add it. But, I'm not sure that we have to validate standard c++ diagnostics because this patch just re-uses them.

Reviewer: Thanks, got it.

> Enabling this macro allows `constant_ptr` to use constant pointers as underlying

Reviewer: ```suggestion
This allows `constant_ptr` to use constant pointers as underlying
```

> pointer types. So, conversions from `constant_ptr` to raw pointers will return

Reviewer: ```suggestion
pointer types. Thus, conversions from `constant_ptr` to raw pointers will return
```

constant pointers and writing to const pointers will be diagnosed by the
front-end.

> This behavior is not following SYCL spec since `constant_ptr` conversions to the

Reviewer: ```suggestion
This behavior is not following the SYCL spec since `constant_ptr` conversions to the
```

> # Overview

Reviewer 2: Why is this file in this folder? Shouldn't this go into e.g. doc/extensions/contant_ptr?

Reviewer 3: According to my understanding this not an extension of the functionality defined by the specification, but rather an implementation detail.

Reviewer: To keep up with the spirit of "implementation details are DPC++", I'd suggest renaming this into `PreprocessorMacros.md`

Author: Sure.

This file describes macros that have effect on SYCL compiler and run-time.

### RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR

The spec assumes that the SYCL implementation does address space deduction. However,
for our implementation, the deduction is performed in the middle end, where it's

> an implementation doing address space deduction in the middle end, where it's

Reviewer: ```suggestion
```

hard to provide user friendly diagnositcs.
Due to these problems writing to raw pointers obtained from `constant_ptr` is not
diagnosed now.
The user can enable diagnostics upon writing to such pointers via enabling the
`RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` macro.

> `RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` preprocessor macro.

Reviewer: ```suggestion
```
